### PR TITLE
fix (dataset editor): allow Source tab readOnly mode

### PR DESCRIPTION
**Pull request by graceguo-supercat**
### SUMMARY
Based on the discussion in #11478, this PR propose add readOnly mode for Source Tab for Dataset Editor.

cc @benceorlai @zuzana-vej 

**Default**: everyone is allowed to edit Source:
LWW0W3HoPJ

**Disabled**: when feature flag `ENABLE_DATASET_SOURCE_EDIT` is disabled, for example, we can disable edit permission for certain users:
Screen Shot 2020-11-23 at 10 39 51 AM

### TEST PLAN
CI and unit tests

### ADDITIONAL INFORMATION

- [X] Has associated issue: #11478
- [X] Changes UI

### CODE REVIEW

**Comment by benceorlai** [comment]
I think the padlock should apply only when the user enters the Dataset edit screen from Explore. When the enter through Data >Datasets>Edit, the padlock should not apply. In the latter case, the user's intent is clear, to edit the dataset, and the padlock can get annoying quickly.

**Comment by graceguo-supercat** [comment]
> When the enter through Data >Datasets>Edit, the padlock should not apply. In the latter case, the user's intent is clear, to edit the dataset, and the padlock can get annoying quickly.

Hi @benceorlai thanks for the feedback, but i could not agree with this point. When a dataset is large and shared by multiple users/teams, many time you want to _**check and make sure**_ the definition of its metrics and calculated columns before actually make chart out of it.

In the explore view context, i would believe the majority of use cases for Datasets > Edit, is to see dataset details instead of actual change it.

File: superset/config.py [hunk 1]
```diff
@@ -300,6 +300,7 @@ def _try_json_readsha(  # pylint: disable=unused-argument
 DEFAULT_FEATURE_FLAGS: Dict[str, bool] = {
     # Experimental feature introducing a client (browser) cache
     "CLIENT_CACHE": False,
+    "ENABLE_DATASET_SOURCE_EDIT": True,
```
Comment (ktmud) [inline superset/config.py:303]: nit: Can we name this `DISABLE_DATASET_SOURCE_EDIT` and set it to `False` by default? IMO every feature flag should be false by default, similar to every A/B testing experience starts with a current base (new feature being disabled) vs a new feature. We set to default feature flag to true only when we put an A/B testing full-on, but they will be cleaned up eventually.

**Comment by benceorlai** [comment]
> > When the enter through Data >Datasets>Edit, the padlock should not apply. In the latter case, the user's intent is clear, to edit the dataset, and the padlock can get annoying quickly.
> 
> Hi @benceorlai thanks for the feedback, but i could not agree with this point. When a dataset is large and shared by multiple users/teams, many time you want to _**check and make sure**_ the definition of its metrics and calculated columns before actually make chart out of it.
> 
> In the explore view context, i would believe the majority of use cases for Datasets > Edit, is to see dataset details instead of actual change it.

That's an interesting point you raise, and then the solution could be the introduction of a new "View" action, in addition to "Edit". My approach to this is to align the user intent and the UX, and I think the "Edit" term is very unambiguous to know what the user should expect. It seems that View (=see details), Edit (=make global changes) and Copy (=make 1-off changes) options would cover every possible intent.

**Comment by Steejay** [comment]
@benceorlai i think the idea of _viewing_ the dataset properties has merits but doesn't necessarily need to be a part of the modal experience itself. if the majority of the use cases are to "see dataset details instead of actually changing it" or (similarly) "check and make sure" the properties are right, there could be an opportunity to expose this info in the side panel _before_ entering the "Edit Dataset" flow. Sim to what we have here but make this view more accessible:


Screen Shot 2020-11-24 at 11 47 39 AM

**Comment by graceguo-supercat** [comment]
here is a real example: a calculated column `Capped duration`: 
Screen Shot 2020-11-24 at 12 40 50 PM

**Comment by zuzana-vej** [comment]
I agree with **edit** vs. **view** in future. In ideal case though, you don't have to go to the modal to view calculated metric's / column's definition. You can hover over it in the new left side bad and see the definition / calculation - the SQL expression basically. For now, I think the current solution is OK, and it's better to be on a safer side.

File: superset/config.py (hunk 1 above)
Comment (graceguo-supercat) [reply superset/config.py]: fixed.

**Comment by ktmud** [comment]
> When a dataset has a lot of calculated columns and metrics, users have to visit this pretty often to make sure the definition.

Not sure if everyone knows this, but you can hover on the little help icon after a pre-defined metric/calculated column to see its definition:

**Comment by ktmud** [comment]
I'm not sure how to do it, but should we place the padlock near the input fields? I'm worried users wouldn't know it's clickable. It would also be nice to have some tooltip or explanatory text on why it is locked.

Or, we can place a "Unlock Edit" link in the warning alert (potentially restyle it to make it more visible).

**Comment by Steejay** [comment]
@ktmud thinking we could position the padlock here w the supporting copy.

Screen Shot 2020-11-25 at 10 26 40 AM
Screen Shot 2020-11-25 at 10 26 47 AM

**Comment by ktmud** [comment]
@Steejay You design looks good!

Here's another alternative:

**Comment by Steejay** [comment]
that could work too @ktmud! i think it would be nice to keep all action buttons in the same region though.

**Comment by graceguo-supercat** [comment]
@ktmud and @Steejay Your design looks good, but please note only `Source` tab has need control. Both header and footer section, are shared with other tabs. We can do hide and show for Padlock if user is in other tab, but I just want to emphasis **"Source" tab** need extra careful protection.

Mostly cases, when user edit Metrics or Columns, they can find error and recover it right away. (Remove existed metrics and columns are dangerous too, you might remove something other ppl use). This also happens pretty often, so we don't want to add extra click for modify Metrics and Columns tab.

**Comment by ktmud** [comment]
The warning text applies to other tabs so I think it makes sense to apply the lock to other tabs, too. As you have noted, deleting a metric/column is also dangerous.

**Comment by Steejay** [comment]
good reference: Mac (System Preferences > Security & Privacy) has a sim padlock experience. Some tabs allow editing some dont unless padlock is unlocked.

**Comment by graceguo-supercat** [comment]
The extra feature flag `DISABLE_DATASET_SOURCE_EDIT`, only meant to disable other user change dataset, but allow teams to enrich dataset. Additional Padlock will make add/modify metric and columns more troublesome, which happens very often in airbnb. 

@ktmud and @Steejay I think for this PR we should follow the agreement from last meetup: https://github.com/apache/incubator-superset/issues/11478#issuecomment-720125272
If you have more options, bring up to meetup and get feedback from more audience.

**Comment by graceguo-supercat** [comment]
Added additional description for the `padlock`:
aqULTDdE1u

Note: i still prefer to keep pad inside `Source` tab and not in the shared area with other tabs, just to make `Source` tab a little special.

**Comment by Steejay** [comment]
i'd recommend adhering to the alignment set by the original context to reduce visual clutter. Suggesting we move the lock to the bottom (but still inside the Source tab) so we dont break up the content of the form by having the padlock in the middle. Attached a mock for reference. 

Frame 213 (2)

**Comment by ktmud** [comment]
> i'd recommend adhering to the alignment set by the original context to reduce visual clutter. Suggesting we move the lock to the bottom (but still inside the Source tab) so we dont break up the content of the form by having the padlock in the middle. Attached a mock for reference.
> 
> Frame 213 (2)

Curious to see how does this look in the virtual datasource tab though? There's a very tall code editor that takes a lot of space.

**Comment by graceguo-supercat** [comment]
move padlock to the bottom of Source tab:
WOaFWC1yzb

The SQL Editor has fixed 25 lines of height.

**Comment by Steejay** [comment]
looks good to me @graceguo-supercat

**Comment by ktmud** [comment]
Sorry for keep nitpicking, but is there a way to add disabled state to the SQL editor as well? People might not know they need to scroll.

**Comment by graceguo-supercat** [comment]
> Sorry for keep nitpicking, but is there a way to add disabled state to the SQL editor as well? People might not know they need to scroll.

it is disabled, you can't edit text in SQL Editor. But you can't tell from screenshot.

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 2]
```diff
@@ -142,4 +147,65 @@ describe('DatasourceEditor', () => {
       wrapper.find(Field).find({ fieldKey: 'fetch_values_predicate' }).exists(),
     ).toBe(true);
   });
+
+  describe('enable edit Source tab', () => {
+    beforeAll(() => {
+      isFeatureEnabledMock = jest
+        .spyOn(featureFlags, 'isFeatureEnabled')
+        .mockImplementation(
+          feature => feature === 'ENABLE_DATASET_SOURCE_EDIT',
```
Comment (ktmud) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:156]: I think this should be updated to `DISABLE_DATASET_SOURCE_EDIT` as well.

Why did the test pass even when the feature flag is not correct?

**Review by ktmud — APPROVED** [review]
LGTM with one nit with tests.

There's probably also room for improvements on the SQL editor but both non-blocking:

1. We could make it shorter so that modal footer can be visible by default. Currently users have to scroll even on very large screens. I don't think the SQL editor needs that much space by default. We can always make it resizable if we want to give users the option to view the full SQL query in one screen.
2. We could probably make the disabled state more visible by giving it the same gray background as the table selectors.

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx (hunk 2 above)
Comment (graceguo-supercat) [reply superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:156]: fixed. The logic here is **_if feature is not enabled_**, though used the wrong feature name, this feature "disable edit", is not enabled.

**Comment by graceguo-supercat** [comment]
for the previous comments:
1) I shorten the Editor height from 24 lines => 20 lines. I don't want to make this really small because in airbnb many virtual dataset are huge...
2) I added grey background for the readOnly mode for Editor. please see the updated UI in description.

**Pull request merged**